### PR TITLE
feat: notify template owner on manual build failures

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
Preview:

Screenshot 2024-08-13 at 15 22 02

Related to https://github.com/coder/team-coconut/issues/16 and it depends on https://github.com/coder/coder/pull/14250

### CODE REVIEW

File: coderd/provisionerdserver/provisionerdserver.go [hunk 1]
```diff
@@ -1585,6 +1615,42 @@ func (s *server) notifyWorkspaceDeleted(ctx context.Context, workspace database.
 	}
 }
 
+func (s *server) notifyTemplateOwnerAboutManualBuildFailure(ctx context.Context, workspace database.Workspace, build database.WorkspaceBuild) {
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver.go:1618]:
```suggestion
func (s *server) notifyTemplateManualBuildFailure(ctx context.Context, workspace database.Workspace, build database.WorkspaceBuild) {
```

File: coderd/provisionerdserver/provisionerdserver.go [hunk 2]
```diff
@@ -1585,6 +1615,42 @@ func (s *server) notifyWorkspaceDeleted(ctx context.Context, workspace database.
...
+		default:
+			reason = string(build.Reason)
+		}
+	} else {
+		reason = string(build.Reason)
+		s.Logger.Warn(ctx, "invalid build reason when sending deletion notification",
+			slog.F("reason", reason), slog.F("workspace_id", workspace.ID), slog.F("build_id", build.ID))
+	}
+
+	if _, err := s.NotificationsEnqueuer.Enqueue(ctx, workspace.OwnerID, notifications.TemplateWorkspaceDeleted,
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver.go:1642]: Workspace deleted?

File: coderd/provisionerdserver/provisionerdserver.go (hunk 1 above)
Comment (dannykopping) [reply coderd/provisionerdserver/provisionerdserver.go:1618]: What's going on with this function? 🤔

File: coderd/provisionerdserver/provisionerdserver_test.go [hunk 3]
```diff
@@ -1804,6 +1804,168 @@ func TestNotifications(t *testing.T) {
 			})
 		}
 	})
+
+	t.Run("TemplateManualBuildFailed", func(t *testing.T) {
+		t.Parallel()
+
+		t.Run("NotifyTemplateOwner", func(t *testing.T) {
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver_test.go:1811]: Please add given/when/then

File: coderd/provisionerdserver/provisionerdserver_test.go [hunk 4]
```diff
@@ -1804,6 +1804,168 @@ func TestNotifications(t *testing.T) {
...
+				Type: &proto.FailedJob_WorkspaceBuild_{
+					WorkspaceBuild: &proto.FailedJob_WorkspaceBuild{
+						State: []byte{},
+					},
+				},
+			})
+			require.NoError(t, err)
+
+			require.Len(t, notifEnq.Sent, 1)
+			require.Equal(t, notifEnq.Sent[0].UserID, templateOwner.ID)
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver_test.go:1885]: Please check for the template name to ensure we're receiving the expected notification.

File: coderd/provisionerdserver/provisionerdserver_test.go [hunk 5]
```diff
@@ -1804,6 +1804,168 @@ func TestNotifications(t *testing.T) {
...
+			notifEnq := &testutil.FakeNotificationsEnqueuer{}
+
+			//	Otherwise `(*Server).FailJob` fails with:
+			// audit log - get build {"error": "sql: no rows in result set"}
+			ignoreLogErrors := true
+			srv, db, ps, pd := setup(t, ignoreLogErrors, &overrides{
+				notificationEnqueuer: notifEnq,
+			})
+
+			initiator := dbgen.User(t, db, database.User{})
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver_test.go:1906]: There's a lot of setup in these tests, and I think there's a bunch in common.
Can we split this out into a function?

File: coderd/provisionerdserver/provisionerdserver_test.go (hunk 3 above)
Comment (dannykopping) [reply coderd/provisionerdserver/provisionerdserver_test.go:1811]: In general, please try make your tests read sensibly and clearly.
Try highlight the part of the test which is significant; you've got a _lot_ of setup here, and you could make it clearer what is expected and why.

Tests are not just validation, they're _communication_. They tell a story about how you expect your code to run.
If it's just large chunks of text then it makes comprehension more difficult, and the intent of the test much harder to discern.

File: coderd/provisionerdserver/provisionerdserver.go [hunk 6]
```diff
@@ -1020,7 +1020,25 @@ func (s *server) FailJob(ctx context.Context, failJob *proto.FailedJob) (*proto.
...
+				return nil, xerrors.Errorf("get template to notify manual build failed: %w", err)
+			}
+
+			owner, err := s.Database.GetUserByID(ctx, workspace.OwnerID)
+			if err != nil {
+				return nil, xerrors.Errorf("get owner to notify manual build failed: %w", err)
+			}
+
+			// Only notify the template creator if the build was initiated by someone else
+			if build.InitiatorID != template.CreatedBy {
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver.go:1038]: Why just the template creator, and not all template admins - for example?

File: coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        title_template,
+        body_template,
+        "group",
+        actions
+    )
+VALUES (
+        '48a9d2b9-3655-430c-a31a-2442479e7519',
+        'Template Manual Build Failure',
+        E'Workspace with template "{{.Labels.name}}" failed to build',
+        E'Hi {{.UserName}}\n\nThe workspace **{{.Labels.workspaceName}}**, using the template **{{.Labels.name}}**, failed during a manual build({{.Labels.transition}}) initiated by the user **{{.Labels.initiator}}**.',
```
[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql:14]: > This sentence is a bit too complex. Can we break this up?

Sure, but we should be cautious about breaking it down too much. In my opinion, having one sentence per line doesn’t always help and can sound unnatural. I’ve been using ChatGPT to generate these copies, but it might be a good idea to get some help from @stirby.

> I think it'd also be good to reference the template version.

Absolutely, that would be helpful.

File: coderd/provisionerdserver/provisionerdserver.go (hunk 6 above)
Comment (BrunoQuaresma) [reply coderd/provisionerdserver/provisionerdserver.go:1038]: I followed the instructions described in issue https://github.com/coder/team-coconut/issues/16, and in my opinion, it makes sense. I believe only the template owner should receive build failure notifications from workspaces built from that template; otherwise, it could become too spammy.

>  manual build failure (receiver: owner)

File: coderd/provisionerdserver/provisionerdserver.go (hunk 1 above)
Comment (BrunoQuaresma) [reply coderd/provisionerdserver/provisionerdserver.go:1618]: This is a good question. I think I may have created it for reference and forgot to remove it.

File: coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql]: @stirby It's okay. I will check how to add lists to the email template. Another thing to note is that you wrote this template with the workspace owner in mind, but this email will be sent to the template owner. The purpose of this notification is to track when a template fails.

We might consider changing the "TemplateWorkspaceAutobuildFailed" template to work for this scenario as well as for automatic build failures. 🤔 They are quite similar, in my opinion, and the copy would only need to change slightly. What do you think, @dannykopping?

[the remark this answers is not in the document]
Comment (stirby) [reply coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql]: >  wrote this template with the workspace owner in mind

My mistake, fixed below. I would bias towards using one notification template for both cases. 

The following copy is better suited for template administrators.

```
"Build Failed: Workspace '{workspace-name}' using template '{template-name}'"

Hi {UserName}

A workspace using the template {template-name} failed to build.
- Template version: {template-version} 
- Workspace: {workspace-name}
- Build type: {build-type}
- Initiated by: {initiator}

You can debug this workspace using the build logs below or contact the deployment administrator. 
```

[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql]: We can use the same template for both manual or automatic build failures, and just pass in a label to specify which or conditionalize the template.

I'd prefer a table over an unordered list as presented above, but I don't feel strongly about it.

> I will check how to add lists to the email template

It's just markdown, so the way @stirby wrote it above will work.

[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql]: > I'd prefer a table over an unordered list as presented above, but I don't feel strongly about it.

I rarely see tables being used in emails to display information.

What do you think about the following alternative?

```
Hi {UserName},

The workspace {workspace-name} using the template {template-name} (version {template-version}) failed to build.

This issue occurred during the {build-type} phase and was initiated by {initiator}.

You can debug this workspace using the build logs below or contact the deployment administrator. 
```

[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql]: What would `{build-type}` be here? `manual` vs `automatic`?

[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql]: It would be the transition. Bad name, sorry!

File: coderd/provisionerdserver/provisionerdserver.go [hunk 8]
```diff
@@ -1020,7 +1020,25 @@ func (s *server) FailJob(ctx context.Context, failJob *proto.FailedJob) (*proto.
 			return nil, err
 		}
 
-		s.notifyWorkspaceBuildFailed(ctx, workspace, build)
+		manualBuild := build.Reason.Valid() && build.Reason == database.BuildReasonInitiator
+		if !manualBuild {
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver.go:1024]: Nit: avoid using negative conditions

File: coderd/provisionerdserver/provisionerdserver_test.go [hunk 9]
```diff
@@ -1804,6 +1803,121 @@ func TestNotifications(t *testing.T) {
 			})
 		}
 	})
+
+	t.Run("TemplateManualBuildFailed", func(t *testing.T) {
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver_test.go:1807]: Nice test!

File: coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (stirby) [reply coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql]: I used bullets instead of natural text because I imagine people switching between tabs and glancing back at the content. Finding it in a sentence is just more effort. A table would be fine too.

It's not important enough to continue refactoring but I don't see how the latest iteration is an improvement.

Screenshot 2024-08-15 at 11 52 43 AM

Screenshot 2024-08-15 at 11 52 17 AM

TLDR, I prefer lists/data points over long sentences, but I'll leave it up to you.

File: coderd/provisionerdserver/provisionerdserver.go (hunk 6 above)
Comment (stirby) [reply coderd/provisionerdserver/provisionerdserver.go:1038]: Many customers have 1-2 templates and 3-5 template administrators. We should send it to all template admins.

File: coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply coderd/database/migrations/000243_notifications_template_manual_build_failure.up.sql]: I personally like how the list looks like 👍 

Screenshot 2024-08-15 at 15 43 17

File: coderd/provisionerdserver/provisionerdserver.go [hunk 10]
```diff
@@ -1020,7 +1020,23 @@ func (s *server) FailJob(ctx context.Context, failJob *proto.FailedJob) (*proto.
 			return nil, err
 		}
 
-		s.notifyWorkspaceBuildFailed(ctx, workspace, build)
+		// Only notify auto build failures
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver.go:1023]: I would rather that this comment explain what classifies as an auto build (i.e. not initiated by a user)

File: coderd/provisionerdserver/provisionerdserver.go [hunk 11]
```diff
@@ -1020,7 +1020,23 @@ func (s *server) FailJob(ctx context.Context, failJob *proto.FailedJob) (*proto.
...
+			s.notifyWorkspaceBuildFailed(ctx, workspace, build, workspace.OwnerID)
+		}
+
+		// Notify template admins including owners
+		admins, err := findTemplateAdmins(ctx, s.Database)
+		if err != nil {
+			s.Logger.Warn(ctx, "failed to find template admins for template build failed notification", slog.Error(err))
+		} else {
+			for _, admin := range admins {
+				if admin.ID != build.InitiatorID && !autoBuild {
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver.go:1035]:
```suggestion
				// Only notify users who did not initiate the action.
				if admin.ID != build.InitiatorID && !autoBuild {
```

File: coderd/provisionerdserver/provisionerdserver.go [hunk 12]
```diff
@@ -1095,17 +1111,11 @@ func (s *server) FailJob(ctx context.Context, failJob *proto.FailedJob) (*proto.
...
 
-func (s *server) notifyWorkspaceBuildFailed(ctx context.Context, workspace database.Workspace, build database.WorkspaceBuild) {
-	var reason string
-	if build.Reason.Valid() && build.Reason == database.BuildReasonInitiator {
-		return // failed workspace build initiated by a user should not notify
-	}
-	reason = string(build.Reason)
-
-	if _, err := s.NotificationsEnqueuer.Enqueue(ctx, workspace.OwnerID, notifications.TemplateWorkspaceAutobuildFailed,
+func (s *server) notifyWorkspaceBuildFailed(ctx context.Context, workspace database.Workspace, build database.WorkspaceBuild, receiverID uuid.UUID) {
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver.go:1114]: I don't understand why you're passing in `receiverID` here. Can you elaborate pls?

File: coderd/provisionerdserver/provisionerdserver.go [hunk 13]
```diff
@@ -1114,6 +1124,43 @@ func (s *server) notifyWorkspaceBuildFailed(ctx context.Context, workspace datab
 	}
 }
 
+func (s *server) notifyTemplateBuildFailed(ctx context.Context, workspace database.Workspace, build database.WorkspaceBuild, receiverID uuid.UUID) {
+	template, err := s.Database.GetTemplateByID(ctx, workspace.TemplateID)
+	if err != nil {
+		s.Logger.Warn(ctx, "failed to get template", slog.Error(err))
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver.go:1130]: Think about how an operator might debug with these logs; how would they know that this is related to sending notifications?

Try think of log lines as communication. You're communicating with an operator who is a) possibly not familiar with the product or less experienced, and b) trying to debug an issue, possibly under stress. Vague log lines are almost as bad as no log lines at all, sometimes worse because they can mislead or distract.

File: coderd/provisionerdserver/provisionerdserver_test.go [hunk 14]
```diff
@@ -1804,6 +1804,155 @@ func TestNotifications(t *testing.T) {
...
+				// Ensure that during automated builds, all template admins and owners
+				// receive notifications.
+				receivers: []uuid.UUID{ownerA.ID, ownerB.ID, tplAdmin.ID},
+			},
+		}
+
+		for _, c := range tc {
+			t.Run(c.name, func(t *testing.T) {
+				// Given: a template and a workspace build
+				isManualBuild := c.reason == database.BuildReasonInitiator
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver_test.go:1863]: Since you have this logic in multiple places, it might be worth creating a helper method.

Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver_test.go:1863]: Try and group your code into logical blocks, and use whitespace for visual assistance. Think of your code as having the same visual hierarchy that you'd design in your frontends.

For example, `isManualBuild` is only needed lower down: group them logically.

You've got a whole lot of code below which is doing several things; add a bit of spaces and/or comments to give the reader's "eye" some space to "breathe", and group related concepts/lines together to infer relationships.

File: coderd/provisionerdserver/provisionerdserver_test.go [hunk 15]
```diff
@@ -1804,6 +1804,155 @@ func TestNotifications(t *testing.T) {
...
+							State: []byte{},
+						},
+					},
+				})
+				require.NoError(t, err)
+
+				// Then: send the template build failed notifications
+				var buildFailedNotifications []*testutil.Notification
+				for _, n := range notifEnq.Sent {
+					if n.TemplateID == notifications.TemplateTemplateBuildFailed {
```
Comment (dannykopping) [inline coderd/provisionerdserver/provisionerdserver_test.go:1930]: What if they're not the template you're expecting?

Your tests should not only prove that something works; it should also prove that it _only_ works in _precisely_ the way you're expecting, and any relevant deviation should fail the test.

**Comment by mtojek** [comment]
@BrunoQuaresma I will take over this PR, not sure if I create a one or modify yours, just FYI 👍

**Comment by BrunoQuaresma** [comment]
> @BrunoQuaresma I will take over this PR, not sure if I create a one or modify yours, just FYI 👍

Sure, whatever you think is best. Just tag me on your PR so I can take a look and learn.

**Comment by mtojek** [comment]
Close in favor of https://github.com/coder/coder/pull/14419
